# 2GB file is downloaded from
# https://www.kaggle.com/datasets/hhs/health-insurance-marketplace?select=Rate.csv 



Rate.csv renamed as health_insurance.csv 

# Reading data with pandas

In [4]:
import pandas as pd

In [5]:
%%time
df = pd.read_csv('health_insurance.csv')
df.head()

Wall time: 23.5 s


BusinessYear StateCode  IssuerId SourceName  VersionNum  \
0          2014        AK     21989       HIOS           6   
1          2014        AK     21989       HIOS           6   
2          2014        AK     21989       HIOS           6   
3          2014        AK     21989       HIOS           6   
4          2014        AK     21989       HIOS           6   

            ImportDate  IssuerId2  FederalTIN RateEffectiveDate  \
0  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
1  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
2  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
3  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
4  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   

  RateExpirationDate  ... IndividualRate IndividualTobaccoRate Couple  \
0         2014-12-31  ...          29.00                   NaN    NaN   
1         2014-12-31  ...          36.95                   NaN   73.9   
2         2014-12-31  ...          36.95                   NaN   73.9   
3         2014-12-31  ...          32.00                   NaN    NaN   
4         2014-12-31  ...          32.00                   NaN    NaN   

  PrimarySubscriberAndOneDependent  PrimarySubscriberAndTwoDependents  \
0                              NaN                                NaN   
1                           107.61                             107.61   
2                           107.61                             107.61   
3                              NaN                                NaN   
4                              NaN                                NaN   

   PrimarySubscriberAndThreeOrMoreDependents  CoupleAndOneDependent  \
0                                        NaN                    NaN   
1                                     107.61                 144.56   
2                                     107.61                 144.56   
3                                        NaN                    NaN   
4                                        NaN                    NaN   

   CoupleAndTwoDependents  CoupleAndThreeOrMoreDependents  RowNumber  
0                     NaN                             NaN         14  
1                  144.56                          144.56         14  
2                  144.56                          144.56         15  
3                     NaN                             NaN         15  
4                     NaN                             NaN         16  

[5 rows x 24 columns]

Reading the 2GB csv file took 23.5 s using pandas

# Reading data with modin and ray

In [12]:
pip install modin

     -------------------------------------- 68.9/68.9 kB 946.3 kB/s eta 0:00:00
     ---------------------------------------- 1.8/1.8 MB 8.3 MB/s eta 0:00:00
     -------------------------------------- 102.6/102.6 kB 5.8 MB/s eta 0:00:00
     -------------------------------------- 126.7/126.7 kB 7.3 MB/s eta 0:00:00
     ---------------------------------------- 2.0/2.0 MB 20.8 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [3]:
import modin.pandas as mpd

In [4]:
#with ray as the computation engine (https://modin.readthedocs.io/en/stable/)

import ray
ray.shutdown()
ray.init(runtime_env={'env_vars': {'__MODIN_AUTOIMPORT_PANDAS__': '1'}})

2023-01-12 13:00:29,387	INFO worker.py:1529 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 


RayContext(dashboard_url='127.0.0.1:8265', python_version='3.9.7', ray_version='2.2.0', ray_commit='b6af0887ee5f2e460202133791ad941a41f15beb', address_info={'node_ip_address': '127.0.0.1', 'raylet_ip_address': '127.0.0.1', 'redis_address': None, 'object_store_address': 'tcp://127.0.0.1:65507', 'raylet_socket_name': 'tcp://127.0.0.1:62575', 'webui_url': '127.0.0.1:8265', 'session_dir': 'C:\\Users\\archa\\AppData\\Local\\Temp\\ray\\session_2023-01-12_13-00-24_836582_22084', 'metrics_export_port': 58897, 'gcs_address': '127.0.0.1:56107', 'address': '127.0.0.1:56107', 'dashboard_agent_listen_port': 52365, 'node_id': '74e81a18e6bd5ec3167544b8de0d28eafa109749be18dcca81fc25a0'})

In [5]:
%%time
mdf = mpd.read_csv('health_insurance.csv')

Wall time: 16.5 s


Reading the 2GB csv file took 16.5 s using modin with ray

# Reading data with dask

In [7]:
# https://docs.dask.org/en/stable/generated/dask.dataframe.read_csv.html
from dask import dataframe as dd

In [8]:
%%time
dask_df = dd.read_csv('health_insurance.csv')

Wall time: 13.8 ms


Dask is better than pandas and modin with computational time of 0.014 s, whereas the other two took 23.5 s and 16.5 s respectively

# Using Dask to read the file and performing ingestion

In [16]:
dask_df = dd.read_csv('health_insurance.csv', delimiter = ',')
dask_df.head()

BusinessYear StateCode  IssuerId SourceName  VersionNum  \
0          2014        AK     21989       HIOS           6   
1          2014        AK     21989       HIOS           6   
2          2014        AK     21989       HIOS           6   
3          2014        AK     21989       HIOS           6   
4          2014        AK     21989       HIOS           6   

            ImportDate  IssuerId2  FederalTIN RateEffectiveDate  \
0  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
1  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
2  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
3  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
4  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   

  RateExpirationDate  ... IndividualRate IndividualTobaccoRate Couple  \
0         2014-12-31  ...          29.00                   NaN    NaN   
1         2014-12-31  ...          36.95                   NaN   73.9   
2         2014-12-31  ...          36.95                   NaN   73.9   
3         2014-12-31  ...          32.00                   NaN    NaN   
4         2014-12-31  ...          32.00                   NaN    NaN   

  PrimarySubscriberAndOneDependent  PrimarySubscriberAndTwoDependents  \
0                              NaN                                NaN   
1                           107.61                             107.61   
2                           107.61                             107.61   
3                              NaN                                NaN   
4                              NaN                                NaN   

   PrimarySubscriberAndThreeOrMoreDependents  CoupleAndOneDependent  \
0                                        NaN                    NaN   
1                                     107.61                 144.56   
2                                     107.61                 144.56   
3                                        NaN                    NaN   
4                                        NaN                    NaN   

   CoupleAndTwoDependents  CoupleAndThreeOrMoreDependents  RowNumber  
0                     NaN                             NaN         14  
1                  144.56                          144.56         14  
2                  144.56                          144.56         15  
3                     NaN                             NaN         15  
4                     NaN                             NaN         16  

[5 rows x 24 columns]

In [13]:
dask_df.info()

<class 'dask.dataframe.core.DataFrame'>
Columns: 24 entries, BusinessYear to RowNumber
dtypes: object(10), float64(9), int64(5)

In [20]:
print("Number of rows: ", len(dask_df.index))
print("Number of columns : {}".format(len(dask_df.columns)))

Number of rows:  12694445
Number of columns : 24


In [14]:
dask_df.columns

Index(['BusinessYear', 'StateCode', 'IssuerId', 'SourceName', 'VersionNum',
       'ImportDate', 'IssuerId2', 'FederalTIN', 'RateEffectiveDate',
       'RateExpirationDate', 'PlanId', 'RatingAreaId', 'Tobacco', 'Age',
       'IndividualRate', 'IndividualTobaccoRate', 'Couple',
       'PrimarySubscriberAndOneDependent', 'PrimarySubscriberAndTwoDependents',
       'PrimarySubscriberAndThreeOrMoreDependents', 'CoupleAndOneDependent',
       'CoupleAndTwoDependents', 'CoupleAndThreeOrMoreDependents',
       'RowNumber'],
      dtype='object')

There are no special characters or white spaces in the column names, however, they can be removed as follows

In [21]:
# remove special character
dask_df.columns=dask_df.columns.str.replace('\W','',regex=True)

#To remove white space from columns
dask_df.columns = dask_df.columns.str.replace(' ', '')

In [22]:
dask_df.columns

Index(['BusinessYear', 'StateCode', 'IssuerId', 'SourceName', 'VersionNum',
       'ImportDate', 'IssuerId2', 'FederalTIN', 'RateEffectiveDate',
       'RateExpirationDate', 'PlanId', 'RatingAreaId', 'Tobacco', 'Age',
       'IndividualRate', 'IndividualTobaccoRate', 'Couple',
       'PrimarySubscriberAndOneDependent', 'PrimarySubscriberAndTwoDependents',
       'PrimarySubscriberAndThreeOrMoreDependents', 'CoupleAndOneDependent',
       'CoupleAndTwoDependents', 'CoupleAndThreeOrMoreDependents',
       'RowNumber'],
      dtype='object')

# Creating utility file

In [23]:
%%writefile utility.py
import logging
import os
import subprocess
import yaml
import pandas as pd
import datetime 
import gc
import re


################
# File Reading #
################

def read_config_file(filepath):
    with open(filepath, 'r') as stream:
        try:
            return yaml.safe_load(stream)
        except yaml.YAMLError as exc:
            logging.error(exc)


def replacer(string, char):
    pattern = char + '{2,}'
    string = re.sub(pattern, char, string) 
    return string

def col_header_val(df,table_config):
    '''
    replace whitespaces in the column
    and standardized column names
    '''
    df.columns = df.columns.str.lower()
    df.columns = df.columns.str.replace('[^\w]','_',regex=True)
    df.columns = list(map(lambda x: x.strip('_'), list(df.columns)))
    df.columns = list(map(lambda x: replacer(x,'_'), list(df.columns)))
    expected_col = list(map(lambda x: x.lower(),  table_config['columns']))
    expected_col.sort()
    df.columns =list(map(lambda x: x.lower(), list(df.columns)))
    df = df.reindex(sorted(df.columns), axis=1)
    if len(df.columns) == len(expected_col) and list(expected_col)  == list(df.columns):
        print("column name and column length validation passed")
        return 1
    else:
        print("column name and column length validation failed")
        mismatched_columns_file = list(set(df.columns).difference(expected_col))
        print("Following File columns are not in the YAML file",mismatched_columns_file)
        missing_YAML_file = list(set(expected_col).difference(df.columns))
        print("Following YAML columns are not in the file uploaded",missing_YAML_file)
        logging.info(f'df columns: {df.columns}')
        logging.info(f'expected columns: {expected_col}')
        return 0

Writing utility.py


In [24]:
%%writefile file.yaml
file_type: csv
dataset_name: testfile
file_name: health_insurance
table_name: edsurv
inbound_delimiter: ","
outbound_delimiter: "|"
skip_leading_rows: 1
columns: 
    - BusinessYear
    - StateCode
    - IssuerId
    - SourceName
    - VersionNum
    - ImportDate
    - IssuerId2
    - FederalTIN
    - RateEffectiveDate
    - RateExpirationDate
    - PlanId
    - RatingAreaId
    - Tobacco
    - Age
    - IndividualRate
    - IndividualTobaccoRate
    - Couple
    - PrimarySubscriberAndOneDependent
    - PrimarySubscriberAndTwoDependents
    - PrimarySubscriberAndThreeOrMoreDependents
    - CoupleAndOneDependent
    - CoupleAndTwoDependents
    - CoupleAndThreeOrMoreDependents
    - RowNumber

Writing file.yaml


In [25]:
# Read config file
import utility as util
config_data = util.read_config_file("file.yaml")
config_data

{'file_type': 'csv',
 'dataset_name': 'testfile',
 'file_name': 'health_insurance',
 'table_name': 'edsurv',
 'inbound_delimiter': ',',
 'outbound_delimiter': '|',
 'skip_leading_rows': 1,
 'columns': ['BusinessYear',
  'StateCode',
  'IssuerId',
  'SourceName',
  'VersionNum',
  'ImportDate',
  'IssuerId2',
  'FederalTIN',
  'RateEffectiveDate',
  'RateExpirationDate',
  'PlanId',
  'RatingAreaId',
  'Tobacco',
  'Age',
  'IndividualRate',
  'IndividualTobaccoRate',
  'Couple',
  'PrimarySubscriberAndOneDependent',
  'PrimarySubscriberAndTwoDependents',
  'PrimarySubscriberAndThreeOrMoreDependents',
  'CoupleAndOneDependent',
  'CoupleAndTwoDependents',
  'CoupleAndThreeOrMoreDependents',
  'RowNumber']}

In [26]:
config_data['file_name']

'health_insurance'

In [30]:
# read the csv file using config file with dask 
file_type = config_data['file_type']
source_file = "./" + config_data['file_name'] + f'.{file_type}'
print(source_file)
df = pd.read_csv(source_file,config_data['inbound_delimiter'])
df.head()

./health_insurance.csv


BusinessYear StateCode  IssuerId SourceName  VersionNum  \
0          2014        AK     21989       HIOS           6   
1          2014        AK     21989       HIOS           6   
2          2014        AK     21989       HIOS           6   
3          2014        AK     21989       HIOS           6   
4          2014        AK     21989       HIOS           6   

            ImportDate  IssuerId2  FederalTIN RateEffectiveDate  \
0  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
1  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
2  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
3  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   
4  2014-03-19 07:06:49      21989  93-0438772        2014-01-01   

  RateExpirationDate  ... IndividualRate IndividualTobaccoRate Couple  \
0         2014-12-31  ...          29.00                   NaN    NaN   
1         2014-12-31  ...          36.95                   NaN   73.9   
2         2014-12-31  ...          36.95                   NaN   73.9   
3         2014-12-31  ...          32.00                   NaN    NaN   
4         2014-12-31  ...          32.00                   NaN    NaN   

  PrimarySubscriberAndOneDependent  PrimarySubscriberAndTwoDependents  \
0                              NaN                                NaN   
1                           107.61                             107.61   
2                           107.61                             107.61   
3                              NaN                                NaN   
4                              NaN                                NaN   

   PrimarySubscriberAndThreeOrMoreDependents  CoupleAndOneDependent  \
0                                        NaN                    NaN   
1                                     107.61                 144.56   
2                                     107.61                 144.56   
3                                        NaN                    NaN   
4                                        NaN                    NaN   

   CoupleAndTwoDependents  CoupleAndThreeOrMoreDependents  RowNumber  
0                     NaN                             NaN         14  
1                  144.56                          144.56         14  
2                  144.56                          144.56         15  
3                     NaN                             NaN         15  
4                     NaN                             NaN         16  

[5 rows x 24 columns]

In [31]:
#validating the header of the file
util.col_header_val(df,config_data)

column name and column length validation passed


1

In [32]:
print("columns of files are:" ,dask_df.columns)
print("columns of YAML are:" ,config_data['columns'])

columns of files are: Index(['BusinessYear', 'StateCode', 'IssuerId', 'SourceName', 'VersionNum',
       'ImportDate', 'IssuerId2', 'FederalTIN', 'RateEffectiveDate',
       'RateExpirationDate', 'PlanId', 'RatingAreaId', 'Tobacco', 'Age',
       'IndividualRate', 'IndividualTobaccoRate', 'Couple',
       'PrimarySubscriberAndOneDependent', 'PrimarySubscriberAndTwoDependents',
       'PrimarySubscriberAndThreeOrMoreDependents', 'CoupleAndOneDependent',
       'CoupleAndTwoDependents', 'CoupleAndThreeOrMoreDependents',
       'RowNumber'],
      dtype='object')
columns of YAML are: ['BusinessYear', 'StateCode', 'IssuerId', 'SourceName', 'VersionNum', 'ImportDate', 'IssuerId2', 'FederalTIN', 'RateEffectiveDate', 'RateExpirationDate', 'PlanId', 'RatingAreaId', 'Tobacco', 'Age', 'IndividualRate', 'IndividualTobaccoRate', 'Couple', 'PrimarySubscriberAndOneDependent', 'PrimarySubscriberAndTwoDependents', 'PrimarySubscriberAndThreeOrMoreDependents', 'CoupleAndOneDependent', 'CoupleAndTwoDepen

# Write the file in pipe separated text file (|) in gz format

In [33]:
dask_df.to_csv("health_insurance.gz", sep = '|', index = False)

In [42]:
#number of files in gz format folder

entries = os.listdir('health_insurance.gz')
len(entries)

31

In [35]:
#size of the gz format folder
import os as os
os.path.getsize('health_insurance.gz')

12288